### PR TITLE
[GISel][RISCV] Add irtranslator/legalizer/selector support for G_FREEZE.

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for G_FREEZE on riscv. It will be selected into a copy instruction.
 
The ll test is copied from the AArch64 patch: https://github.com/llvm/llvm-project/commit/665da596854bf07ee25f368855156dde43845013.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll:29]: Avoid undef here, and non-constant inputs are generally more useful tests

File: llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll
Comment (michaelmaitland) [inline llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: What is the reason to remove `freeze_struct`, `freeze_anonstruct`, and `freeze_anonstruct2`

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: IRTranslator doesn't support passing structs by registers.

> LLVM ERROR: unable to lower arguments: ptr (in function: freeze_anonstruct)
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
      Program arguments: bin/llc -mtriple=riscv32 -mattr=+f,+m,+v --global-isel -o - test.ll
      Running pass 'Function Pass Manager' on module 'test.ll'.
      Running pass 'IRTranslator' on function '@freeze_anonstruct'

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll
Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: > Could just load from memory instead

I still have trouble in legalizing G_FREEZE with fixed-length vectors :(

File: llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll
Comment (aemerson) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: I'll leave this to RISCV maintainers to decide, but IMO it's probably better to have fallbacks enabled in the test so that you get some test output that you can fix in future?

Alternatively you can have a test that just keeps track of unimplemented fallbacks so you don't lose track of these, see `AArch64/GlobalISel/arm64-fallback.ll` as an example

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: Fixed.

File: llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll
Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: From the code, it looks like you are only legalizing with scalable vectors. Maybe we should test with scalable vectors and not fixed vectors?

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/freeze.ll]: Fixed.

**Review by michaelmaitland — APPROVED** [review]
LGTM

**Pull request merged**
